### PR TITLE
allow configurable battleground graphics

### PR DESCRIPTION
**Pull request by nullkiller**
- [x] Add custom terrain patches with custom battlegrounds
- [x] Allow custom battlegrounds for any object
- [x] Allow custom bonuses for terrain patches

### CODE REVIEW

File: lib/Terrain.h [hunk 1]
```diff
@@ -42,6 +42,11 @@ class DLL_LINKAGE BattleField
 	{
 		h & name;
 	}
+
+	bool isSpecial() const
+	{
+		return name.find('_') >= 0; // hack for special battlefields, move to JSON
```
Comment (Nordsoft91) [inline lib/Terrain.h:48]: Are you going to have parameter instead?

Comment (nullkiller) [reply lib/Terrain.h]: yes, already done

File: client/battle/CBattleInterface.cpp [hunk 2]
```diff
@@ -200,15 +201,9 @@ CBattleInterface::CBattleInterface(const CCreatureSet *army1, const CCreatureSet
...
-		auto bfieldType = curInt->cb->battleGetBattlefieldType();
-		if(!vstd::contains(graphics->battleBacks, bfieldType))
-		{
-			logGlobal->error("%s is not valid battlefield type!", static_cast<std::string>(bfieldType));
-		}
-		else
-		{
-			background = BitmapHandler::loadBitmap(graphics->battleBacks[bfieldType], false);
-		}
+		auto bfieldType = curInt->cb->battleGetBattlefieldType().getInfo();
```
Comment (Nordsoft91) [inline client/battle/CBattleInterface.cpp:204]: I'd suggest to check fieldType for null

File: lib/BattleFieldHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ *
+ */
+#include "StdInc.h"
+
+#include <vcmi/Entity.h>
+#include "BattleFieldHandler.h"
+
+BattleFieldInfo * BattleFieldHandler::loadFromJson(const std::string & scope, const JsonNode & json, const std::string & identifier, size_t index)
+{
+	BattleFieldInfo * info = new BattleFieldInfo(BattleField(index), identifier);
```
Comment (Nordsoft91) [inline lib/BattleFieldHandler.cpp:17]: who will delete it?

File: lib/BattleFieldHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+std::vector<JsonNode> BattleFieldHandler::loadLegacyData(size_t dataSize)
+{
+	return std::vector<JsonNode>();
+}
+
+const std::vector<std::string> & BattleFieldHandler::getTypeNames() const
+{
+	static const  std::vector<std::string> types = std::vector<std::string> { "battlefield" };
```
Comment (Nordsoft91) [inline lib/BattleFieldHandler.cpp:70]: extra space

File: test/erm/ERM_BU.cpp [hunk 5]
```diff
@@ -174,7 +174,7 @@ TEST_F(ERM_BU_G, Get2)
 	source << "!?PI;" << std::endl;
 	source << "!!BU:G?v1;" << std::endl;
 
-	EXPECT_CALL(binfoMock, battleGetBattlefieldType()).WillOnce(Return(BattleField("evil_fog")));
+	EXPECT_CALL(binfoMock, battleGetBattlefieldType()).WillOnce(Return(BattleField(5)));
```
Comment (Nordsoft91) [inline test/erm/ERM_BU.cpp:177]: Could you please add a comment what 5 does mean?

File: test/erm/ERM_BU.cpp [hunk 6]
```diff
@@ -155,7 +155,7 @@ TEST_F(ERM_BU_G, Get)
 	source << "!?PI;" << std::endl;
 	source << "!!BU:G?v1;" << std::endl;
 
-	EXPECT_CALL(binfoMock, battleGetBattlefieldType()).WillOnce(Return(BattleField("snow_trees")));
+	EXPECT_CALL(binfoMock, battleGetBattlefieldType()).WillOnce(Return(BattleField(5)));
```
Comment (Nordsoft91) [inline test/erm/ERM_BU.cpp:158]: Could you please add a comment what 5 does mean?

**Comment by nullkiller** [comment]
All unit tests have been fixed locally! Btw we do not run them on CI because they are actually not unit but kind of integration ones and require to boot VLC and other static things unfortunately. And thus they need original game data as well

File: client/battle/CBattleInterface.cpp (hunk 2 above)
Comment (nullkiller) [reply client/battle/CBattleInterface.cpp:204]: It will write an error to log about unresolved id.

**Comment by Alucard648** [comment]
What about having multiple battlefields, chosen at random, for same terrain/object?

**Comment by nullkiller** [comment]
> What about having multiple battlefields, chosen at random, for same terrain/object?

It is more about graphics probably. We can allow battlefield to have multiple graphics and to chose one at random for instance

**Pull request merged**